Simplenote/src/androidTest/java/com/automattic/simplenote/sync/SyncTests.java

```diff
@@ -0,0 +1,696 @@
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeoutException;
+
+import tools.fastlane.screengrab.Screengrab;
+import tools.fastlane.screengrab.UiAutomatorScreenshotStrategy;
+import tools.fastlane.screengrab.cleanstatusbar.CleanStatusBar;
+import tools.fastlane.screengrab.locale.LocaleTestRule;
+
+@LargeTest
+@RunWith(AndroidJUnit4.class)
+public class SyncTests {
+    private static final int LAUNCH_TIMEOUT = 5000;
+
+    private static final String NOTE_FOR_EDITOR_SHOT_TITLE = "Lemon Cake & Blueberry";
+    private static final String NOTE_FOR_INTERLINKING_SHOT_TITLE = "# Colors";
+    private static final String NOTE_FOR_INTERLINKED_NOTE_SHOT_TITLE = "Blueberry Recipes";
+
+    @ClassRule
+    public static final LocaleTestRule localeTestRule = new LocaleTestRule();
+
+    @Rule
+    public ActivityTestRule<NotesActivity> mActivityTestRule = new ActivityTestRule<>(NotesActivity.class);
+
+    @BeforeClass
+    public static void beforeAll() {
+        Screengrab.setDefaultScreenshotStrategy(new UiAutomatorScreenshotStrategy());
+
+        new CleanStatusBar()
+                .setBatteryLevel(100)
+                .setClock("1231")
+                .setWifiVisibility(SHOW)
+                .setBarsMode(TRANSPARENT)
+                .enable();
+    }
+
+    @Test
+    public void publishNoteTest() throws InterruptedException {
+        // Pre-checks if the state is dirty
+        enterThenDisablePasscodeIfNeeded();
+        dismissVerifyEmailScreenIfNeeded();
+        logoutIfNeeded();
+
+        // Login
+        login();
+
+        // Waiting for the notes
+        Thread.sleep(2000);
+        dismissVerifyEmailScreenIfNeeded();
+
+        waitForViewMatching(withText(R.string.all_notes), 500);
+
+        // Creating one
+        addNoteFromNoteList("The title", "The content");
+
+        // Deleting it
+        deleteNoteFromNoteList("The title");
+    }
+
+    @AfterClass
+    public static void afterAll() {
+        CleanStatusBar.disable();
+    }
+
+    // Flows
+
+    private void addNoteFromNoteList(String title, String content) {
+        getViewById(R.id.fab_button).perform(click());
+
+        waitForViewMatching(withId(R.id.note_content), 5000);
+
+        getViewById(R.id.note_content)
+                .perform(click())
+                .perform(typeText(title + "\n\n" + content))
+                .perform(ViewActions.closeSoftKeyboard());
+
+        onView(withContentDescription("More Options")).perform(click());
+
+        onView(withText("Publish")).perform(click());
+
+        onView(withContentDescription("Navigate up")).perform(click());
+
+        waitForViewMatching(withText(R.string.all_notes), 500);
+
+        waitForViewMatching(allOf(withId(R.id.note_title), withText(title)), 5000);
+    }
```

Comment:
It looks like this method does more than its name implies. For such a scenario, I would have the code broken up into three methods instead (naming is just an example):

- `createNoteFromNoteList(...)`
- `publishCurrentNote()`
- `leaveNoteEditor()`

This would allow having more granular tests and steps, which you can still freely combine into some mega-methods.

One of the things that happens once the note is published is generation of a URL, which can be used to open the published note in the browser. So, automating this test would definitely require checking that the URL is generated and can be opened in the browser. Unfortunately, note publishing does not work in the dev-DB, and I forgot to let you know about this earlier 🤦. Sorry for this confusion.

Reply:
I'll split it up. Thanks for the URL generation, I didn't know that.